test/e2e/storage/testsuites/multivolume.go

```diff
@@ -329,6 +330,106 @@ func (t *multiVolumeTestSuite) DefineTests(driver storageframework.TestDriver, p
 			l.config.ClientNodeSelection, resource.Pvc, numPods, true /* sameNode */, false /* readOnly */)
 	})
 
+	// This tests below configuration:
+	// [pod1]           [pod2]
+	// [        node1        ]
+	//   |                 |     <- same volume mode
+	// [volume1]   ->  [restored volume1 snapshot]
+	ginkgo.It("should concurrently access the volume and restored snapshot from pods on the same node [LinuxOnly][Feature:VolumeSnapshotDataSource][Feature:VolumeSourceXFS]", func() {
+		init()
+		defer cleanup()
+
+		if !l.driver.GetDriverInfo().Capabilities[storageframework.CapSnapshotDataSource] {
+			e2eskipper.Skipf("Driver %q does not support volume snapshots - skipping", dInfo.Name)
+		}
+		if pattern.SnapshotType == "" {
+			e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
+		}
+
+		// Create a volume
+		testVolumeSizeRange := t.GetTestSuiteInfo().SupportedSizeRange
+		resource := storageframework.CreateVolumeResource(l.driver, l.config, pattern, testVolumeSizeRange)
+		l.resources = append(l.resources, resource)
+		pvcs := []*v1.PersistentVolumeClaim{resource.Pvc}
+
+		// Create snapshot of it
+		sDriver, ok := driver.(storageframework.SnapshottableTestDriver)
+		if !ok {
+			framework.Failf("Driver %q has CapSnapshotDataSource but does not implement SnapshottableTestDriver", dInfo.Name)
+		}
+		testConfig := storageframework.ConvertTestConfig(l.config)
+		dc := l.config.Framework.DynamicClient
+		dataSource, cleanupFunc := prepareSnapshotDataSourceForProvisioning(f, testConfig, l.config, pattern, l.cs, dc, resource.Pvc, resource.Sc, sDriver, pattern.VolMode, "injected content")
+		defer cleanupFunc()
+
+		// Create 2nd PVC for testing
+		pvc2 := &v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resource.Pvc.Name + "-restored",
+				Namespace: resource.Pvc.Namespace,
+			},
+		}
+		resource.Pvc.Spec.DeepCopyInto(&pvc2.Spec)
+		pvc2.Spec.VolumeName = ""
+		pvc2.Spec.DataSource = dataSource
+
+		pvc2, err := l.cs.CoreV1().PersistentVolumeClaims(pvc2.Namespace).Create(context.TODO(), pvc2, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		pvcs = append(pvcs, pvc2)
+		defer func() {
+			l.cs.CoreV1().PersistentVolumeClaims(pvc2.Namespace).Delete(context.TODO(), pvc2.Name, metav1.DeleteOptions{})
+		}()
+
+		// Test access to both volumes on the same node.
+		TestConcurrentAccessToRelatedVolumes(l.config.Framework, l.cs, l.ns.Name,
```

Comment:
would it be possible to add a check here to make sure that both snapshots have the same contents before calling `TestConcurrentAccessToRelatedVolumes`? I'm not sure if that's doable without spinning up some pods though

Reply:
It is, however, volume prepared using `preparePVCDataSourceForProvisioning()` / `prepareSnapshotDataSourceForProvisioning` is not "compatible" with volume checks used in `multivolume.go` (`utils.CheckReadFromPath` + `utils.CheckWriteToPath`). It would be cool to merge them, but I gave up after an hour, it requires bigger refactoring.

I added code that checks if the volumes have the right content, however, it works only for filesystem volumes. And it does not check that the volumes are "independent", i.e. write to one does not affect the other one - that would be a nice test too, but not with this framework.